background-processor/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+use lightning::chain;
+use lightning::chain::chaininterface::{BroadcasterInterface, FeeEstimator};
+use lightning::chain::keysinterface::{ChannelKeys, KeysInterface};
+use lightning::ln::channelmanager::ChannelManager;
+use lightning::{log_internal, log_trace};
+use lightning::util::logger::Logger;
+use lightning::util::ser::Writeable;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::thread;
+use std::thread::JoinHandle;
+use std::time::{Duration, Instant};
+
+/// BackgroundProcessor takes care of tasks that (1) need to happen periodically to keep
+/// Rust-Lightning running properly, and (2) either can or should be run in the background. Its
+/// responsibilities are:
+/// * Monitoring whether the ChannelManager needs to be re-persisted to disk, and if so,
+///   writing it to disk/backups by invoking the callback given to it at startup.
+///   ChannelManager persistence should be done in the background.
+/// * Calling `ChannelManager::timer_chan_freshness_every_min()` every minute (can be done in the
+///   background).
+///
+/// Note that if ChannelManager persistence fails and the persisted manager becomes out-of-date,
+/// then there is a risk of channels force-closing on startup when the manager realizes it's
+/// outdated. However, as long as `ChannelMonitor` backups are sound, no funds besides those used
+/// for unilateral chain closure fees are at risk.
+pub struct BackgroundProcessor {
+	stop_thread: Arc<AtomicBool>,
+	/// May be used to retrieve and handle the error if `BackgroundProcessor`'s thread
+	/// panics.
+	pub thread_handle: JoinHandle<()>
+}
+
+// Re-export these modules for the logging macros.
+mod util {
+	pub(crate) mod logger {
+		pub(crate) use lightning::util::logger::Record;
+	}
+}
+
+#[cfg(not(test))]
+const CHAN_FRESHNESS_TIMER: u64 = 60;
+#[cfg(test)]
+const CHAN_FRESHNESS_TIMER: u64 = 1;
+
+impl BackgroundProcessor {
+	/// Start a background thread that takes care of responsibilities enumerated in the top-level
+	/// documentation. Marked as `must_use` because otherwise the result is dropped immediately,
+	/// resulting in the thread being terminated.
```

Comment:
I think the thread is allowed to run free in this case? We don't have a custom `Drop` and docs for `JoinHandle` indicate thread runs free on drop.

Reply:
Weird, I thought I ran into a problem necessitating this but I can't reproduce it now. Removed.